hazelcast/src/main/java/com/hazelcast/sql/HazelcastSqlException.java

```diff
@@ -49,7 +56,6 @@ public HazelcastSqlException(
     /**
      * Gets ID of the member that caused or initiated an error condition.
      */
-    @Nonnull
```

Comment:
I don't like the current state of exception cloning. Instead of calling the constructor with signature `(String message, Throwable cause)` by reflection, we can introduce a `recreate()` method in `HazelcastException` with a default implementation and override it in `HazelcastSqlException`, `QueryException`.

Reply:
@burakgok AFAIK we create an exception on the client side having only information from https://github.com/hazelcast/hazelcast/blob/8f68566e49a8c83200c129de29b0af132a9668fe/hazelcast/src/main/java/com/hazelcast/client/impl/protocol/exception/ErrorHolder.java#L23-L26

How `recreate()` could help and be implemented?

Reply:
Let's leave the refactoring to another time.

What I had in mind is if the `HazelcastSqlException` and `QueryException` are meant to be propagated to the client using this mechanism. We would lose some info from the exception (hence removing the `@Nonnull` here) - it seems wrong to me.

Reply:
> removing the @Nonnull here - it seems wrong to me.

Agree, we would force users to think if the return value is null where it's null only for recreated exception (at least for now).

 Maybe after all we should leave `UndefinedErrorCodeException` for these exceptions? 

https://github.com/hazelcast/hazelcast/blob/14ef0dbefa4a6fa00a50a8873aeea02ec6c3f5e4/hazelcast/src/main/java/com/hazelcast/client/UndefinedErrorCodeException.java#L23-L24

Reply:
@ldziedziul I have just realized that the default implementation will use reflection as well. However, even in that case, the `recreate()` method will *correctly* clone all exceptions (if you override it for problematic subclasses, of course).

Reply:
I'm not too happy about it but I think it's OK if we make it nullable. I'm not a fan of non-text information in exceptions in general. Exceptions should be simple, they should be only text+cause+stack trace. If there's anything else, especially if it doesn't appear in the message, it's often lost. Also the exception API design should be focused on specific actions that you can take based on the specific information you can extract extract. I don't see any action that the client could take based on the knowledge of the initiator UUID - would it restart it? That doesn't make sense. The client would just print the error, and the initiator UUID is not even contained in the to-string output. So that's why I don't mind too much if this information is lost.

Btw, technically, it's a breaking change, so we should mention it in the release notes.